File: python/algo/leecode/algo1537.py
from typing import List

# 滑动窗口与双指针 - 双序列双指针 - 4.1 双指针
class Solution:
    def maxSum(self, nums1, nums2) -> int:
        sum1, sum2, inx1, inx2 = 0, 0, 0, 0
        n1, n2 = len(nums1), len(nums2)
        
        while inx1 < n1 or inx2 < n2:
            if inx1 < n1 and inx2 < n2:
                if nums1[inx1] < nums2[inx2]:
                    sum1 += nums1[inx1]
                    inx1 += 1
                elif nums1[inx1] > nums2[inx2]:
                    sum2 += nums2[inx2]
                    inx2 += 1
                else:
                    # equals
                    sum1 += nums1[inx1]
                    inx1 += 1
                    sum2 += nums2[inx2]
                    inx2 += 1
                    sum1 = max(sum1, sum2) % (10 ** 9 + 7)
                    sum2 = sum1
            elif inx1 < n1:
                while inx1 < n1:
                    sum1 += nums1[inx1]
                    inx1 += 1
            else:
                while inx2 < n2:
                    sum2 += nums2[inx2]
                    inx2 += 1
        return max(sum1, sum2) % (10 ** 9 + 7)

    def maxSum2(self, nums1: List[int], nums2: List[int]) -> int:
        # 复习，简化上述代码
        sum1 = sum2 = inx1 = inx2 = 0
        n1, n2 = len(nums1), len(nums2)
        
        while inx1 < n1 and inx2 < n2:
            if nums1[inx1] < nums2[inx2]:
                sum1 += nums1[inx1]
                inx1 += 1
            elif nums1[inx1] > nums2[inx2]:
                sum2 += nums2[inx2]
                inx2 += 1
            else:      # equals
                sum2 = sum1 = max(sum1, sum2) + nums1[inx1]
                inx1 += 1
                inx2 += 1
        for i in range(inx1, n1): sum1 += nums1[i]
        for i in range(inx2, n2): sum2 += nums2[i]

        return max(sum1, sum2) % (10 ** 9 + 7)    

if __name__ == "__main__":
    sol = Solution()
    nums1 = [2,4,5,8,10]
    nums2 = [4,6,8,9]
    print(sol.maxSum(nums1, nums2))

File: cplus/leetcode/202511/algo1537.cpp
#include <bits/stdc++.h>
using namespace std;

// 滑动窗口与双指针 - 四、双序列双指针 - 4.1 双指针
class Solution {
public:
    int maxSum(vector<int>& nums1, vector<int>& nums2) {
        // 看错题目了，动态规划 :
        // f[i] = f[i-1] + nums1[i] ,nums1[i] < nums2[j]
        // g[j] = g[j-1] + nums2[j] ,nums1[i] > nums2[j]
        // f[i] = g[j] = max(f[i-1], g[j-1]) + nums1[i] ,nums1[i] == nums2[j]
        const int mod = 1e9 + 7;

        int inx1 = 0, n = nums1.size(), inx2 = 0, m = nums2.size();
        long long f = 0, g = 0;
        while (inx1 < n && inx2 < m) {
            if (nums1[inx1] == nums2[inx2]) {
                f = g = max(f, g) + nums1[inx1];
                inx1 ++, inx2 ++;
            } else if (nums1[inx1] < nums2[inx2]) 
                f += nums1[inx1++];
            else g += nums2[inx2++];
        }
        for (int i=inx1; i < n; i++) f +=nums1[i];
        for (int i=inx2; i < m; i++) g +=nums2[i];

        return max(f, g) % mod;  
    }
};